### PR TITLE
Tile SDPA flash attention over KV dimension via explicit Python for-loop

### PR DESCRIPTION
**Pull request by ani300**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] feature

#### What this PR does:

Replaces the `spyre_hint(tiles={"max_seqlen_kv": ...})` context in
`spyre__sdpa_overrideable` (`torch_spyre/_inductor/decompositions.py`) with an
explicit Python `for` loop over KV blocks of 64 — the classic flash-attention
outer loop over KV blocks.

Because the SDPA decomp runs during `torch.compile` tracing and the loop bound
(`num_kv_blocks = ceil(max_seqlen_kv / 64)`) is a Python int, the loop unrolls
at trace time into N statically-sliced sub-graphs that write sequentially into
the same `M` / `denominator` / `output` accumulators via `torch.ops.spyre.copy_f`
(an in-place op). Each iteration slices `key`/`value` (and the causal mask /
`attn_bias`) along the KV axis, computes partial scores against the full query,
and folds the block into the online-softmax accumulators. The other tiling hints
(`batch_size`, `num_heads`, `max_seqlen_q`, `work_div`) are retained.

Also:
- Adds `mha_prefill_kv_tail(_causal)` SDPA test cases with `max_seqlen_kv=130`
  to exercise the short final KV block (non-multiple-of-64 tail).
- Shrinks the existing SDPA prefill test sequence lengths (256 → 128) and decode
  KV (257 → 129) to cut compile/run time; block size in the decomp is unchanged.

#### Special notes for your reviewer:

- The online-softmax fold is copied verbatim from the pre-change body; only the
  KV extent becomes a per-block slice. At a whole number of 64-blocks the result
  is numerically identical to the prior hinted version (verified ~5e-7 vs. a
  full-softmax reference; tail block at kv=130 verified separately).
- The per-block causal slice `causal_mask[..., :, start:end]` is correct because
  PyTorch 2.13's `F.scaled_dot_product_attention(is_causal=True)` uses top-left
  causal alignment, matching `spyre::causal_mask`'s `tril`.
- `mha_decode`/`gqa_decode` remain `expect_fail` — their failure is pre-existing
  and unrelated to KV tiling (malformed 5-tuple param sets / non-transposed
  decode inputs), so it is out of scope here.
- Design spec and implementation plan are included under
  `docs/superpowers/`.

#### Does this PR introduce a user-facing change?

No. This changes the internal Spyre SDPA lowering strategy; the op's signature,
outputs, and numerics are unchanged.

#### Additional note:

Block size is fixed at 64 (the fp16 stick size). Hint re-tuning and tiling of
any axis other than `max_seqlen_kv` are explicitly out of scope.

### CODE REVIEW

File: torch_spyre/_inductor/decompositions.py [hunk 1]
```diff
@@ -413,6 +413,7 @@ def spyre__sdpa_overrideable(
 
     kv_block_size = 64
     q_block_size = 64
```
Comment (cyang49) [inline torch_spyre/_inductor/decompositions.py:476]: should block size be configurable?

File: torch_spyre/_inductor/decompositions.py [hunk 2]
```diff
@@ -453,26 +454,32 @@ def spyre__sdpa_overrideable(
                 tiles={"max_seqlen_q": max(1, max_seqlen_q // q_block_size)}
             ):
                 with spyre_hint(
-                    tiles={"max_seqlen_kv": max(1, max_seqlen_kv // kv_block_size)}
+                    work_div={"num_heads": 4, "max_seqlen_q": 8, "max_seqlen_kv": 8}
```
Comment (cyang49) [inline torch_spyre/_inductor/decompositions.py]: seqlen work_div hints will likely stop working in new loop tiling structures --> can verify by looking at logs

File: torch_spyre/_inductor/decompositions.py (hunk 1 above)
Comment (ani300) [reply torch_spyre/_inductor/decompositions.py:476]: probably, just trying to get it to work first

**Comment by mudhakar** [comment]
I'm closing #3674 in favor of this PR. Three things worth carrying over.

1. Block counts > 6 did not work. Can you do a quick check. 8k at block size of 64 is 128 blocks. PR's tests with 2-3 blocks are just fine. 
2. with KV sliced in graph and only slice's consumer named, H-tiling was silently wrong (tile 0 is correct, rest is wrong). Did you check of numerical accuracy. 
3. Measured 8k configs, both exact (0.00% vs eager-CPU):

| phase | shape | chunks | tiling | compile |
|---|---|---|---|---|
| prefill | `B=1 H=8 Lq=512 Lk=8192 D=128` | 4 × 2048 | H:4 + Lq:2 | 46 s |
| decode | `B=4 H=8 Lq=1 Lk=8192 D=128` | 4 × 2048 | H:4 | 28 s |

**Comment by ani300** [comment]
**Why H-tiling can't be ported from #3674 yet**

@mudhakar @tardieu — following up on the note about carrying #3674's H-tiling over here — we tried, and it's blocked one layer down. Recording why.

Activating the `num_heads` tile in the decomp (rename the head-axis seed `"_h"` → `"num_heads"`) compiles cleanly but is numerically wrong: **head tile 0 is correct, every later head tile is ~85% wrong** (26.8% aggregate mismatch), whenever `num_kv_blocks >= 2` — which every prefill case here hits by construction. This is exactly the "tile 0 correct, rest wrong" symptom flagged at #3674 close-out. The 1-KV-block case passes on all heads, so the tile is fine in isolation; the break is a per-tile *offset* into K/V, not arithmetic.

**Root cause is in WSR, not the decomp.** The K/V blocks are sliced *in-graph* here, so their reads are intermediate buffers. The only in-graph naming primitive, `spyre_hint(named_dims=...)`, seeds an op's **output** and `continue`s (`propagate_named_dims.py:416-456`) — it never runs the backward input analysis (`_compute_named_dims`, `:473-474`), so the upstream K/V read never gets a head-dim binding, and no `_untracked` warning fires (silent). Only **graph inputs** get buffer-level naming that every read inherits (`:406-408`).

That's the whole reason #3674 is exact and we can't lift it as-is: **#3674's K/V chunks are graph inputs** (sliced by the caller, outside the traced region, named via `name_tensor_dims`). `spyre__sdpa_overrideable` is a decomposition with an upstream-fixed `(query, key, value, ...)` signature whose body is traced whole — its K/V slices are intermediate buffers, never inputs, and there's no in-graph primitive to name an intermediate buffer's dims wholesale. (`Lq` tiling works precisely because the block slice varies along *seq*, not *head*, so it never needs a per-tile offset into K/V.)

**So porting H-tiling needs a WSR-layer change, not a decomp edit.** Leading options: (A) add a primitive to seed an intermediate buffer's `_dim_prop_info` at buffer level — the general fix, also unblocks batch tiling; (B) fold heads into the batch axis so the split lands on a named leading axis; (C) an HOP/nested-compile boundary that reproduces #3674's caller. Leaning A, deciding now.

The `"_h"` → `"num_heads"` rename is **not committed** — reverting restores the known-good baseline (`14218a4`) exactly. Full writeup: `docs/superpowers/specs/2026-08-12-sdpa-head-tile-blocker-readout.md`.

**Comment by ani300** [comment]
#4259 superseded this
